### PR TITLE
Read Only Fields

### PR DESCRIPTION
**Pull request by lithorias**
**What**
A new property was added to Salesforce syncable allowing users to specify a list of read only fields that would be defined in Salesforce.

**Why**
We are seeing problems when syncing read only fields from the model to Salesforce. This allows us to define a set of readonly fields that will only be synced from Salesforce to the model and not the other way around.

**Summary of Changes**
- Added / exposed a new field to salesforce sync for salesforce syncable models
- Added additional check when figuring out what fields to update to ignore readonly fields
- Updated version
- Updated Readme
- Updated specs

**Additional Notes**
- The rails5 branch is included in this so we can create a release out of that whole thing and close all those branches.

### CODE REVIEW

**Review by rob-meh — APPROVED** [review]
I don't really know what this is, but the syntax looks good 😄

File: lib/salesforce_ar_sync/salesforce_sync.rb [hunk 1]
```diff
@@ -122,101 +128,96 @@ def salesforce_attributes_to_set(attributes = {})
...
         self.class.salesforce_sync_attribute_mapping.each do |key, value|
-          if self.respond_to?(value)
+          next unless respond_to?(value)
 
-            #Checkboxes in SFDC Cannot be nil.  Here we check for boolean field type and set nil values to be false
-            attribute_value = self.send(value)
-            if is_boolean?(value) && attribute_value.nil?
-              attribute_value = false
-            end
+          # Checkboxes in SFDC Cannot be nil.  Here we check for boolean field type and set nil values to be false
```
Comment (michaelrychly) [inline lib/salesforce_ar_sync/salesforce_sync.rb:168]:
```suggestion
          # Checkboxes in SFDC Cannot be nil. Here we check for boolean field type and set nil values to be false
```

File: spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb [hunk 2]
```diff
@@ -22,30 +30,32 @@ def email_address_changed?
...
+  FirstName: 'Rose',
+  LastName: 'Gonzalez',
+  Phone: '(512) 757-6000',
+  SystemModstamp: '2012-03-26T19:54:50.000Z',
+  NumberOfPosts__c: 2,
+  WebId__c: 68_836
 }
 
 # this message is missing the Phone field on purpose
 # Salesforce does not include empty fields  in the body
```
Comment (michaelrychly) [inline spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb:50]:
```suggestion
# Salesforce does not include empty fields in the body
```

File: spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb [hunk 3]
```diff
@@ -160,17 +172,16 @@ class TestSyncable < ActiveRecord::Base
...
   describe '.salesforce_id_attribute_name' do
-    it "returns the salesforce Id attribute name" do
+    it 'returns the salesforce Id attribute name' do
       expect(Contact.salesforce_id_attribute_name).to eq(:Id)
     end
   end
 
   describe '.salesforce_sync_web_id?' do
-    it "should default to false" do
+    it 'should default to false' do
```
Comment (michaelrychly) [inline spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb:184]:
```suggestion
    it 'defaults to false' do
```

File: spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb [hunk 4]
```diff
@@ -182,30 +193,30 @@ class TestSyncable < ActiveRecord::Base
   end
 
   describe '.salesforce_default_attributes_for_create' do
-    it "should default to an empty hash" do
+    it 'should default to an empty hash' do
```
Comment (michaelrychly) [inline spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb:196]:
```suggestion
    it 'defaults to an empty hash' do
```

File: spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb [hunk 5]
```diff
@@ -220,7 +231,7 @@ def custom_name
 
     context 'sync_inbound_delete set to true' do
       it 'should return true' do
```
Comment (michaelrychly) [inline spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb:233]:
```suggestion
      it 'returns true' do
```

File: spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb [hunk 6]
```diff
@@ -256,7 +267,7 @@ def sync_delete
 
     context 'sync_outbound_delete set to false' do
       it 'should return false' do
```
Comment (michaelrychly) [inline spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb:269]:
```suggestion
      it 'returns false' do
```

File: spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb [hunk 7]
```diff
@@ -290,17 +301,17 @@ def sync_delete
 
     context 'when SYNC_ENABLED is false in the global SalesforceArSync.config hash' do
       it 'should return true' do
```
Comment (michaelrychly) [inline spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb:303]:
```suggestion
      it 'returns true' do
```

File: spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb [hunk 8]
```diff
@@ -403,24 +413,23 @@ def custom_sync
       expect(@hash[:phone_number]).to eq(sample_outbound_message_hash[:Phone])
     end
 
-    it 'hash should only include 5 keys' do
-      expect(@hash.keys.count).to eq(5)
+    it 'hash should only include 6 keys' do
```
Comment (michaelrychly) [inline spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb:416]:
```suggestion
    it 'only include 6 keys in the hash' do
```

File: spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb [hunk 9]
```diff
@@ -403,24 +413,23 @@ def custom_sync
...
 
       contact.salesforce_attributes_to_set(sample_outbound_message_hash).tap do |hash|
         expect(hash[:first_name]).to be_nil
       end
     end
-
   end
 
   describe '#salesforce_process_update' do
     it 'should update and save record with values passed in to hash' do
```
Comment (michaelrychly) [inline spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb:431]:
```suggestion
    it 'updates and saves record with values passed in to hash' do
```

File: spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb [hunk 10]
```diff
@@ -433,7 +442,7 @@ def custom_sync
     end
 
     it 'should nil out any values not specified in the message from Salesforce' do
```
Comment (michaelrychly) [inline spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb:444]:
```suggestion
    it 'nils out any values not specified in the message from Salesforce' do
```

**Review by michaelrychly — APPROVED** [review]
All 'should' references could be eliminated in:
spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb

**Review by jhaig — DISMISSED** [review]

File: spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb [hunk 11]
```diff
@@ -370,57 +384,55 @@ class SyncTest < ActiveRecord::Base
...
   end
 
   describe '#salesforce_attributes_to_set' do
-
     before(:each) do
       @hash = Contact.new.salesforce_attributes_to_set(sample_outbound_message_hash)
     end
 
-    it 'hash should include salesforce_updated_at' do
+    it 'hash includes salesforce_updated_at' do
```
Comment (michaelrychly) [inline spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb:405]:
```suggestion
    it 'includes salesforce_updated_at' do
```

File: spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb [hunk 12]
```diff
@@ -370,57 +384,55 @@ class SyncTest < ActiveRecord::Base
...
       expect(@hash[:salesforce_updated_at]).to eq(sample_outbound_message_hash[:SystemModstamp])
     end
 
-    it 'hash should include salesforce_id' do
+    it 'hash includes salesforce_id' do
```
Comment (michaelrychly) [inline spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb:409]:
```suggestion
    it 'includes salesforce_id' do
```

File: spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb [hunk 13]
```diff
@@ -370,57 +384,55 @@ class SyncTest < ActiveRecord::Base
...
       expect(@hash[:salesforce_id]).to eq(sample_outbound_message_hash[:Id])
     end
 
-    it 'hash should only include keys and values in the mapping that we have setters for' do
+    it 'hash only includes keys and values in the mapping that we have setters for' do
```
Comment (michaelrychly) [inline spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb:413]:
```suggestion
    it 'only includes keys and values in the mapping that we have setters for' do
```

File: spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb [hunk 14]
```diff
@@ -370,57 +384,55 @@ class SyncTest < ActiveRecord::Base
...
       expect(@hash[:first_name]).to eq(sample_outbound_message_hash[:FirstName])
       expect(@hash[:last_name]).to eq(sample_outbound_message_hash[:LastName])
       expect(@hash[:phone_number]).to eq(sample_outbound_message_hash[:Phone])
     end
 
-    it 'hash should only include 5 keys' do
-      expect(@hash.keys.count).to eq(5)
+    it 'hash only includes 6 keys' do
```
Comment (michaelrychly) [inline spec/salesforce_ar_sync/salesforce_ar_sync_spec.rb:419]:
```suggestion
    it 'only includes 6 keys' do
```

**Review by michaelrychly — APPROVED** [review]

**Review by rob-meh — DISMISSED** [review]

**Review by jhaig — DISMISSED** [review]

**Review by rob-meh — APPROVED** [review]

**Review by jhaig — APPROVED** [review]

**Pull request merged**